### PR TITLE
feat: Hardware API

### PR DESCRIPTION
**Pull request by vansangpfiev**
## Describe Your Changes

```
curl --request GET --url http://127.0.0.1:39281/v1/hardware
```
```json
{
    "cpu": {
        "arch": "amd64",
        "cores": 8,
        "instructions": [
            "fpu",
            "mmx",
            "sse",
            "sse2",
            "sse3",
            "ssse3",
            "sse4_1",
            "sse4_2",
            "pclmulqdq",
            "avx",
            "avx2",
            "aes",
            "f16c"
        ],
        "model": "AMD Ryzen Threadripper PRO 5955WX 16-Cores"
    },
    "gpus": [
        {
            "additional_information": {
                "compute_cap": "8.6",
                "driver_version": "535.183"
            },
            "free_vram": 24246,
            "name": "NVIDIA GeForce RTX 3090",
            "total_vram": 24576,
            "version": "12.2"
        }
    ],
    "os": {
        "name": "Ubuntu 24.04.1 LTS",
        "version": "24.04.1 LTS (Noble Numbat)"
    },
    "power": {
        "battery_life": 0,
        "charging_status": "",
        "is_power_saving": false
    },
    "ram": {
        "available": 14335725568,
        "total": 16768770048,
        "type": ""
    },
    "storage": {
        "available": 0,
        "total": 0,
        "type": ""
    }
}
```
- Model usage estimation will be in a separate PR
## Fixes Issues

- https://github.com/janhq/cortex.cpp/issues/1591
- https://github.com/janhq/cortex.cpp/issues/1613
- https://github.com/janhq/cortex.cpp/issues/1603

## Self Checklist

- [ ] Added relevant comments, esp in complex areas
- [ ] Updated docs (for bug fixes / features)
- [ ] Created issues for follow-up changes or refactoring needed

### CODE REVIEW

File: engine/utils/hardware/os_info.h [hunk 1]
```diff
@@ -0,0 +1,33 @@
+#pragma once
+#include <json/json.h>
+#include <string>
+#include "hwinfo/hwinfo.h"
+
+namespace hardware {
```
Comment (namchuai) [inline engine/utils/hardware/os_info.h]: I think we better move those structs into `common` folder.

File: engine/utils/cortex_utils.h [hunk 2]
```diff
@@ -31,206 +24,9 @@ inline std::string logs_folder = "./logs";
...
-      std::cout << resetColor << c;
-      colorIndex = 0;
-    } else {
-      std::cout << rainbowColors[colorIndex % 2] << c;
-      colorIndex++;
-    }
-  }
-
-  std::cout << resetColor;  // Reset color at the endreturn;
+inline std::string rtrim(const std::string& str) {
```
Comment (namchuai) [inline engine/utils/cortex_utils.h]: maybe we can remove this function or move it to `string_utils`

File: engine/services/model_service.cc [hunk 3]
```diff
@@ -659,17 +661,112 @@ cpp::result<bool, std::string> ModelService::StartModel(
 #undef ASSIGN_IF_PRESENT
 
     CTL_INF(json_data.toStyledString());
+    // Calculate ram/vram needed to load model
```
Comment (namchuai) [inline engine/services/model_service.cc:665]: Should we separate into another function?

File: engine/e2e-test/test_cli_engine_list.py [hunk 4]
```diff
@@ -9,6 +9,7 @@ class TestCliEngineList:
     @pytest.fixture(autouse=True)
     def setup_and_teardown(self):
         # Setup
+        stop_server()
```
Comment (namchuai) [inline engine/e2e-test/test_cli_engine_list.py:12]: Hmm, I think we should make every test in isolation, which mean the test which start the server will have to stop the server itself.

File: engine/common/engine_servicei.h [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+  Json::Value ToJson() const {
+    Json::Value root;
+    root["name"] = name;
+    root["version"] = version;
+    root["engine"] = engine;
+    return root;
+  }
+};
+
+class EngineServiceI {
```
Comment (namchuai) [inline engine/common/engine_servicei.h:37]: Hmm, I'm not sure if we need this interface.

File: engine/e2e-test/test_cli_engine_list.py (hunk 4 above)
Comment (vansangpfiev) [reply engine/e2e-test/test_cli_engine_list.py:12]: I agree. But some race condition in the tests make `stop_server` is not called. I cannot reproduce in the local machine, so make the e2e pass first

File: engine/common/engine_servicei.h (hunk 5 above)
Comment (vansangpfiev) [reply engine/common/engine_servicei.h:37]: I would like to follow SOLID principle, the `HardwareService` and `ModelService` class does not need to access all function of `EngineService`

**Review by namchuai — APPROVED** [review]
✅ Overall lgtm. Only some nits and should not be blocked. Great job!

File: engine/services/model_service.cc (hunk 3 above)
Comment (vansangpfiev) [reply engine/services/model_service.cc:665]: Will add TODO and do it in another PR. Still have the model estimation need to be merged.

File: engine/utils/hardware/os_info.h (hunk 1 above)
Comment (vansangpfiev) [reply engine/utils/hardware/os_info.h]: Updated

File: engine/utils/cortex_utils.h (hunk 2 above)
Comment (vansangpfiev) [reply engine/utils/cortex_utils.h]: updated

**Pull request merged**